File: js/untyped/State.js

function go(call = true) {
  this.cell.value.state = this;
  if (call) this.cell.value.doAction();
  else return [this.cell.value, 'doAction'];
}

function toString() {
  return this.name;
}

function State(name) {
  if (new.target) throw new TypeError();
  const block = {
    name,
    toString,
    cell: null,
    go
  };
  return block;
}

State.init = function(block, action0, action1) {
  const actions = {
    action0,
    action1
  };
  Object.assign(block, actions);
  Object.seal(block);
}

module.exports = State;

File: c/typed/state.c

#include <iostream>

#include "State.h"
#include "Cell.h"
#include "Action.h"
#include "Value.h"

State::State(const char* n, Action& a0, Action& a1) :
  name(n), action0(&a0), action1(&a1) {}

void State::go() {
  cell->value->state = this;
  cell->value->do_action();
}

std::ostream& operator<<(std::ostream& os, const State* state) {
  os << state->name;
  return os;
}
